segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ChunkStorageManager.java

```diff
@@ -0,0 +1,1435 @@
+            // Which means target should now point to it as next after concat is complete.
+
+            // If there are chunks that can be appended together then concat them.
+            if (chunksToConcat.size() > 1) {
+                // Concat
+
+                ConcatArgument[] concatArgs = new ConcatArgument[chunksToConcat.size()];
+                for (int i = 0; i < chunksToConcat.size(); i++) {
+                    concatArgs[i] = ConcatArgument.fromChunkInfo(chunksToConcat.get(i));
+                }
+
+                if (!useAppend && chunkStorage.supportsConcat()) {
+                    int length = chunkStorage.concat(concatArgs);
+                } else {
+                    concatUsingAppend(concatArgs);
+                }
+
+                // Set the pointers
+                target.setLength(targetSizeAfterConcat);
+                target.setNextChunk(nextChunkName);
+
+                // If target is the last chunk after this then update metadata accordingly
+                if (null == nextChunkName) {
+                    segmentMetadata.setLastChunk(target.getName());
+                    segmentMetadata.setLastChunkStartOffset(segmentMetadata.getLength() - target.getLength());
+                }
+
+                // Update metadata for affected chunks.
+                for (int i = 1; i < concatArgs.length; i++) {
+                    txn.delete(concatArgs[i].getName());
+                    segmentMetadata.decrementChunkCount();
+                }
+                txn.update(target);
+                txn.update(segmentMetadata);
+            }
+
+            // Move on to next place in list where we can concat if we are done with append based concats.
+            if (!useAppend) {
+                targetChunkName = nextChunkName;
+            }
+
+            // Toggle
+            useAppend = !useAppend;
+        }
+
+        // Make sure no invariants are broken.
+        segmentMetadata.checkInvariants();
+    }
+
+    private void concatUsingAppend(ConcatArgument[] concatArgs) throws ChunkStorageException {
+        long writeAtOffset = concatArgs[0].getLength();
+        val writeHandle = ChunkHandle.writeHandle(concatArgs[0].getName());
+        for (int i = 1; i < concatArgs.length; i++) {
+            int readAtOffset = 0;
+            val arg = concatArgs[i];
+            int bytesToRead = Math.toIntExact(arg.getLength());
+
+            while (bytesToRead > 0) {
+                byte[] buffer = new byte[Math.min(config.getMaxBufferSizeForChunkDataTransfer(), bytesToRead)];
+                int size = chunkStorage.read(ChunkHandle.readHandle(arg.getName()), readAtOffset, buffer.length, buffer, 0);
+                bytesToRead -= size;
+                readAtOffset += size;
+                writeAtOffset += chunkStorage.write(writeHandle, writeAtOffset, size, new ByteArrayInputStream(buffer, 0, size));
```

Comment:
What if the storage system supports append, but not write at offset?

Reply:
Then it should throw exception.

Reply:
Both ECS and FileSystem can support overwriting old data.
HDFS is truncating the chunks at end before concat. (like below)

However as a general contract I need to handle this case. 

```
 public int doConcat(ConcatArgument[] chunks) throws ChunkStorageException, UnsupportedOperationException {
        ensureInitializedAndNotClosed();
        int length = 0;
        try {
            val sources = new Path[chunks.length - 1];
            for (int i = 1; i < chunks.length; i++) {
                val chunkLength = chunks[i].getLength();
                this.fileSystem.truncate(getFilePath(chunks[i].getName()), chunkLength);
                sources[i - 1] = getFilePath(chunks[i].getName());
                length += chunkLength;
            }
            // Concat source file into target.
            this.fileSystem.concat(getFilePath(chunks[0].getName()), sources);
        } catch (IOException e) {
            throwException(chunks[0].getName(), "doConcat", e);
        }
        return length;
    }
```